<h1>Pandas apply functions faster using Swifter</h1>

<h4>Find out the time required for applying a function to dataframe</h4>

In [ ]:
# Build a dataframe with 10M rows and two columns with values selected randomly between 1 and 1000

In [1]:
import pandas as pd
import numpy as np
import time

In [2]:
df = pd.DataFrame({'X':np.random.randint(1000, size=10000000),'Y':np.random.randint(1000, size=10000000)})

In [3]:
# Function to add the sum of squares of each column

In [4]:
def add_squares(df):
    return df.X**2+df.Y**2

In [5]:
# Check the time it takes to apply the function

In [6]:
start_time = time.time()
df['add_squares']=df.apply(add_squares,axis=1)
print(time.time()-start_time)

309.784300327301


In [ ]:
# It took average around 309.78 secs per loop to apply add_squares method on this dataframe of 100M rows and 2 columns

<h4>Use Swifter Library & Time IT</h4>

In [3]:
import swifter

In [4]:
def add_squares(a,b):
    return a**2+b**2

In [5]:
start_time = time.time()
df ['add_sq']=df.swifter.apply(lambda row:add_squares(row.X,row.Y),axis=1)
print(time.time()-start_time)


71.2806544303894


In [ ]:
# Swift took average around 71.28 secs per loop to apply add_squares method on this dataframe of 100M rows and 2 columns

<h1>RESULTS</h1>

<h4>100k</h4>
-------------                       


3.010223388671875          
6.556478500366211


#FOR 100k rows apply without swift was faster
    
<h4>1M</h4>
-------------

30.70429754257202  
11.413851499557495         


#FOR ONE MILLION rows apply using swift was 3x faster
    
<h4>10M</h4>     
-------------              


#####  TRIAL -1


***            
303.05524255564                
67.80521702766418          

#FOR ONE MILLION rows apply using swift was 5x faster        
    
##### TRIAL -2
*** 
309.784300327301                                                                                                           
71.2806544303894
    
    
##### FOR LOOP (10 M) W/O SWIFT
***     
306.90898180007935   
309.29545187950134   
308.1306691169739  
308.52621626853943   
307.5656614303589   

##### FOR LOOP (10 M) USING SWIFT
***
67.83769202232361                                                                 
69.81787896156311                                      
70.81935858726501                               
71.49488592147827                         
70.08375406265259                                


In [2]:
#CHANGE NO OF ROWS ABOVE TO NOTICE DIFFERENCE 